### ISSUE TITLE
Change uid and gid columns meaning to User SID and Group SID on Windows

### ISSUE THREAD

**Issue by Smjert** [osquery/osquery#7417]
# Blueprint

## The problem

On Windows the concept of a user id or group id that's just a single number and uniquely identifies an entity on that machine, as on Unix platforms, doesn't really exists.
Windows uses a SID, which is a string composed by `S` prefix and 4 numeric components separated by a dash: a revision level, an identifier authority value, a domain identifier and a relative identifier (or RID). For example `S-1-5-32-544`.

Currently all the tables that contain a `uid` or `gid` column and are implemented on Windows are using only the RID part of the SID to emulate a value that Unix like.
Not only showing this data alone loses information, but Windows API's do not work using that part alone, they need to have the full SID to do queries/filtering.

This brings us to a couple of the issues we have today:
1. To partially resolve the limitation of the `gid` or `uid`, since we can't use them in Windows API, some of the tables (like the `users` and the `groups` table), have to add a platform specific column to also contain the respective user SID (currently `uuid` on the `users` table) and group SID (`group_sid` on the `groups` table).
This causes though the need to create platform specific queries for Windows, which have to use those columns instead of `uid` or `gid`, beyond the current inconsistencies with naming, due to a separate Apple platform need to also represent a user with an `uuid`.

2. Currently the `users` table does not contain a column with the `group_sid` column, so if you need to join the `users` table with the `groups` one, then you are forced to do so with the `gid` column.
This would be fine if the `gid` was a SID on Windows, but since it's not, what happens is that when processing a `SELECT * FROM users as u JOIN groups as g ON u.gid=g.gid` query, sqlite will call the `groups` table once for each user, and the `groups` table is forced to return all the groups existing on the machine, without being able to filter.
This causes osquery to do more work then needed, but also causes additional overhead to the Windows LSASS process, because it too has to do the work of retrieving the whole list of groups on the machine, for every user present in the `users` table.
This becomes particularly pathological if osquery is installed on a Windows domain controller.

3. Even when using the `group_sid` column from `groups` table in a query, the table still doesn't implement any filtering; it will always return the full list.

## The proposal

A part from the performance issues, numeric `uid` and `gid` on Windows really don't mean much, so what I'm proposing is that we should change the meaning of `uid` and `gid` columns for Windows so that they are both SIDs; this also means changing the type of the column to TEXT.
This has the benefit that although there are multiple tables to change, the majority of the existing queries would not need to change.
Join queries especially will be written the same on all the platforms.

Tables that need changes:
`users`: `uid` and `gid` column types become TEXT and represent SIDs on Windows. The `uuid` column becomes unused by Windows and should be moved to the extended schema part for Apple.
`groups`: `gid` column type becomes TEXT and `group_sid` gets removed, since it's substituted by `gid`.
`user_groups`: `uid` and `gid` column types become TEXT and represent SIDs on Windows.
`processes`: `uid` and `gid`  column types become TEXT and represent SIDs on Windows.
`file`: `uid` and `gid`  column types become TEXT and represent SIDs on Windows.

**Comment by Smjert** [comment]
One bonus point is that, specifically for the `users` table, I realized that there are couple of other things to change/that will change.

Currently the table `users` and the table `groups` are meant to list **local** users and **local** groups.
The meaning of **local** here means users that have been created on the local machine, instead of a remote machine, an Active Directory.
In the `users` implementation we have a couple of fallbacks when we cannot retrieve a `gid` by listing it via group APIs; the first uses to the primary group id as taken from a user structure, which gives only the RID part and it can be a local or domain group. The problem here is that if we failed to get the group via API, most likely it's a domain group and it's the built-in `Domain Users`, which is not listed anywhere because technically it's not a local group.

The second fallback uses the RID part of the user SID. This for instance happens with the SYSTEM "user", which doesn't even have a user structure to get the primary group from. I don't think this is correct.

So in short, for those two cases I would currently actually return an empty SID, because what we are showing now is either not a group or not a **local** group and it both cases those values cannot be joined anyway with the `groups` table.

**Comment by directionless** [comment]
We talked about this in office hours on 2021-12-21. I think consensus was (though this is probably seen through my bias)

It seems good to have these values be correct for the OS (eg: SIDs as proposed). And it seems totally reasonable to add appropriate columns. This might address the performance impact, though it would require platform specific join operations

Changing the existing columns is interesting -- I think the correctness and simplicity is appealing. **But** this represents a schema change that would impact consumers of this data, _and_ it likely effects far more tables than anticipated. (`grep -Er '\"(u|g)id' specs/ | grep -v -E 'posix|linux|darwin' | awk -F: '{print $1}' | uniq` shows 12).

There seems to be support for changing the schemas in an osquery 6, for sometime in early 2022. But I don't think we settled on whether to aim for a unified schema, or different schemas per platform.

Whether or not a command line flag allowing someone to opt in would help the transition or not is unclear.
